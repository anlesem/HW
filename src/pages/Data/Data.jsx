import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { getDataAllThunk } from '../../store/data/actions';
import { getDataError, getDataData, getDataLoad } from '../../store/data/selectors';

import style from './Data.module.scss';

import Button from '@mui/material/Button';
import { ProgressBar } from '../../components/ProgressBar/ProgressBar';

export const Data = () => {
  const data = useSelector(getDataData);
  const load = useSelector(getDataLoad);
  const error = useSelector(getDataError);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getDataAllThunk());
  }, []);

  return (
    <div className={style.wrap}>
      {load && <ProgressBar />}
      {error && (
        <div className={style.error}>
          <h2 className={style.text}> Ошибка получения данных</h2>
          <Button
            variant="outlined"
            data-testid="data-button-reload"
            onClick={() => dispatch(getDataAllThunk())}>
            Попробовать снова
          </Button>
        </div>
      )}
      <ul className={style.list}>
        {data.map((item, idx) => {
          return (
            <li key={idx} className={style.item} data-testid={'data-item-' + idx}>
              <h2 className={style.title}>{item.title}</h2>
              <p>{item.body}</p>
            </li>
          );
        })}
      </ul>
    </div>
  );
};
