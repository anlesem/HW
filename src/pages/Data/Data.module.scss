.wrap {
  position: relative;
  width: 100%;
  height: 100%;
}

.error {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  text-align: center;
}

.text {
  font-weight: bold;
  font-size: 20px;
  margin-bottom: 40px;
}

.list {
  width: 100%;
  height: 100%;
  padding: 40px 0;
  overflow-y: auto;
}

.item {
  padding: 10px 40px;
}

.title {
  font-weight: bold;
  font-size: 16px;
}
