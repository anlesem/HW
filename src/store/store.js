import { createStore, combineReducers, compose, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import { profileReducer } from './profile/reducer';
import { chatsReducer } from './chats/reducer';
import { messagesReducer } from './messages/reducer';
import { dataReducer } from './data/reducer';

const persistConfig = {
  key: 'root',
  storage,
  blacklist: ['data']
};

const rootReducers = combineReducers({
  profile: profileReducer,
  chats: chatsReducer,
  messages: messagesReducer,
  data: dataReducer
});

const persistedReducers = persistReducer(persistConfig, rootReducers);

export const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const store = createStore(persistedReducers, composeEnhancers(applyMiddleware(thunk)));

export const persistor = persistStore(store);
