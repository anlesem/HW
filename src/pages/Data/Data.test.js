import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import { Data } from './Data';

import { Provider } from 'react-redux';
import { store } from '../../store/store';

describe('Data', () => {
  it('Компонент существует', () => {
    expect(Data).toBeInstanceOf(Function);
  });

  it('Снимок состояния', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Routes>
            <Route path="data" element={<Data />} />
          </Routes>
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment(<Data />)).toMatchSnapshot();
  });

  it('Состояние при старте', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Routes>
            <Route path="data" element={<Data />} />
          </Routes>
        </BrowserRouter>
      </Provider>
    );

    setTimeout(() => {
      expect(screen.getByTestId('data-item-1')).toBeTruthy();
    }, 2000);
  });
});
