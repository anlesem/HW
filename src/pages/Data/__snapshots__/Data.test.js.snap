// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Снимок состояния 1`] = `<DocumentFragment />`;
